### ISSUE TITLE
Hyperopt requiring indicator_space() method.

### ISSUE THREAD

**Issue by Heraklaitos** [freqtrade/freqtrade#5480]
## Describe your environment

  * Operating system: Ubuntu 21.04 (with VMware virtual machine)
  * Python Version: 3.9.5
  * CCXT version: 1.55.28

  * Freqtrade Version: develop-8a9407ba

  
## Your question

I have been trying to optimize some of my parameters in my sample strategy. As far as I see with freqtrade 2021.4 we no longer need to create an explicit hyperopt file. By reviewing the example code in https://www.freqtrade.io/en/stable/hyperopt/ I have created a sample strategy given below. 

```
import numpy as np  # noqa
import pandas as pd  # noqa
from freqtrade.optimize.space import SKDecimal, Dimension, Integer
from pandas import DataFrame
from functools import reduce

from typing import Any, Callable, Dict, List

from freqtrade.strategy import (BooleanParameter, CategoricalParameter, DecimalParameter,
                                IStrategy, IntParameter)

import talib.abstract as ta
import freqtrade.vendor.qtpylib.indicators as qtpylib


class MyFirstStrategy(IStrategy):

    # Strategy interface version - allow new iterations of the strategy interface.
    # Check the documentation or the Sample strategy to get the latest version.
    INTERFACE_VERSION = 2

    # Minimal ROI designed for the strategy.
    # This attribute will be overridden if the config file contains "minimal_roi".
    minimal_roi = {
        "0": 0.06337,
        "27": 0.03385,
        "86": 0.01406,
        "120": 0
    }


    # Optimal stoploss designed for the strategy.
    # This attribute will be overridden if the config file contains "stoploss".
    stoploss = -0.49122


    # Trailing stoploss
    trailing_stop = False
    # trailing_only_offset_is_reached = False
    # trailing_stop_positive = 0.01
    # trailing_stop_positive_offset = 0.0  # Disabled / not configured

    # Optimal timeframe for the strategy.
    timeframe = '1h'

    # Run "populate_indicators()" only for new candle.
    process_only_new_candles = False

    # These values can be overridden in the "ask_strategy" section in the config.
    use_sell_signal = True
    sell_profit_only = False
    ignore_roi_if_buy_signal = False

    # Number of candles the strategy requires before producing valid signals
    startup_candle_count: int = 30

    # Optional order type mapping.
    order_types = {
        'buy': 'limit',
        'sell': 'limit',
        'stoploss': 'limit',
        'stoploss_on_exchange': False
    }

    # Optional order time in force.
    order_time_in_force = {
        'buy': 'gtc',
        'sell': 'gtc'
    }
    
    plot_config = {
        # Main plot indicators (Moving averages, ...)
        'main_plot': {
            'tema': {},
            'sar': {'color': 'white'},
        },
        'subplots': {
            # Subplots - each dict defines one additional plot
            "MACD": {
                'macd': {'color': 'blue'},
                'macdsignal': {'color': 'orange'},
            },
            "RSI": {
                'rsi': {'color': 'red'},
            }
        }
    }
    # BUY PARAMS
    buy_rsi = IntParameter(10, 50, default=30, space='buy')
    buy_rsi_enabled = CategoricalParameter([True, False], default=True, space='buy')
    buy_trigger = CategoricalParameter(['buy_bb_lowerband1', 'buy_bb_lowerband2', 'buy_bb_lowerband3', 'buy_bb_lowerband4'],
                                       default='buy_bb_lowerband2', space='buy')

    # SELL PARAMS
    sell_rsi = IntParameter(60, 90, default=30, space='sell')
    sell_rsi_enabled = CategoricalParameter([True, False], default=True, space='sell')
    sell_trigger = CategoricalParameter(['sell_bb_middle1', 'sell_bb_upper1',
                                         'sell_bb_middle2', 'sell_bb_upper2',
                                         'sell_bb_middle3', 'sell_bb_upper3',
                                         'sell_bb_middle4', 'sell_bb_upper4'], default='sell-bb_middle2', space='sell')


    def populate_indicators(self, dataframe: DataFrame, metadata: dict) -> DataFrame:
        """
        Adds several different TA indicators to the given DataFrame

        Performance Note: For the best performance be frugal on the number of indicators
        you are using. Let uncomment only the indicator you are using in your strategies
        or your hyperopt configuration, otherwise you will waste your memory and CPU usage.
        :param dataframe: Dataframe with data from the exchange
        :param metadata: Additional information, like the currently traded pair
        :return: a Dataframe with all mandatory indicators for the strategies
        """
        
        dataframe['buy_rsi'] = ta.RSI(dataframe)
        dataframe['sell_rsi'] = ta.RSI(dataframe)

        # Overlap Studies
        # ------------------------------------

        # Bollinger Bands
        bollinger1 = qtpylib.bollinger_bands(qtpylib.typical_price(dataframe), window=20, stds=1)
        dataframe['buy_bb_lowerband1'] = bollinger1['lower']
        dataframe['buy_bb_middleband1'] = bollinger1['mid']
        dataframe['buy_bb_upperband1'] = bollinger1['upper']

        bollinger2 = qtpylib.bollinger_bands(qtpylib.typical_price(dataframe), window=20, stds=2)
        dataframe['buy_bb_lowerband2'] = bollinger2['lower']
        dataframe['buy_bb_middleband2'] = bollinger2['mid']
        dataframe['buy_bb_upperband2'] = bollinger2['upper']

        bollinger3 = qtpylib.bollinger_bands(qtpylib.typical_price(dataframe), window=20, stds=3)
        dataframe['buy_bb_lowerband3'] = bollinger3['lower']
        dataframe['buy_bb_middleband3'] = bollinger3['mid']
        dataframe['buy_bb_upperband3'] = bollinger3['upper']

        bollinger4 = qtpylib.bollinger_bands(qtpylib.typical_price(dataframe), window=20, stds=4)
        dataframe['buy_bb_lowerband4'] = bollinger4['lower']
        dataframe['buy_bb_middleband4'] = bollinger4['mid']
        dataframe['buy_bb_upperband4'] = bollinger4['upper']

        return dataframe

    def populate_buy_trend(self, dataframe: DataFrame, metadata: dict) -> DataFrame:
        """
        Buy strategy Hyperopt will build and use.
        """
        conditions = []

        # GUARDS AND TRENDS
        if self.buy_rsi_enabled.value:
            conditions.append(dataframe['buy_rsi'] < self.buy_rsi.value)

        # TRIGGERS
        if self.buy_trigger.value:
            if self.buy_trigger.value == 'buy_bb_lowerband1':
                conditions.append(dataframe['close'] < dataframe['buy_bb_lowerband1'])
            if self.buy_trigger.value == 'buy_bb_lowerband2':
                conditions.append(dataframe['close'] < dataframe['buy_bb_lowerband2'])
            if self.buy_trigger.value == 'buy_bb_lowerband3':
                conditions.append(dataframe['close'] < dataframe['buy_bb_lowerband3'])
            if self.buy_trigger.value == 'buy_bb_lowerband4':
                conditions.append(dataframe['close'] < dataframe['buy_bb_lowerband4'])

        # Check that the candle had volume
        conditions.append(dataframe['volume'] > 0)

        if conditions:
            dataframe.loc[
                reduce(lambda x, y: x & y, conditions),
                'buy'] = 1

        return dataframe

    def populate_sell_trend(self, dataframe: DataFrame, metadata: dict) -> DataFrame:
        """
        Based on TA indicators, populates the sell signal for the given dataframe
        :param dataframe: DataFrame populated with indicators
        :param metadata: Additional information, like the currently traded pair
        :return: DataFrame with buy column
        """
        conditions = []

        # GUARDS AND TRENDS
        if self.sell_rsi_enabled.value:
            conditions.append(dataframe['rsi'] > self.sell_rsi.value)

        # TRIGGERS
        if self.sell_trigger.value:
            if self.sell_trigger.value == 'sell_bb_middle1':
                conditions.append(dataframe['close'] < dataframe['sell_bb_middle1'])
            if self.sell_trigger.value == 'sell_bb_middle2':
                conditions.append(dataframe['close'] < dataframe['sell_bb_middle2'])
            if self.sell_trigger.value == 'sell_bb_middle3':
                conditions.append(dataframe['close'] < dataframe['sell_bb_middle3'])
            if self.sell_trigger.value == 'sell_bb_middle4':
                conditions.append(dataframe['close'] < dataframe['sell_bb_middle4'])
            if self.sell_trigger.value == 'sell_bb_upper1':
                conditions.append(dataframe['close'] < dataframe['sell_bb_upper1'])
            if self.sell_trigger.value == 'sell_bb_upper2':
                conditions.append(dataframe['close'] < dataframe['sell_bb_upper2'])
            if self.sell_trigger.value == 'sell_bb_upper3':
                conditions.append(dataframe['close'] < dataframe['sell_bb_upper3'])
            if self.sell_trigger.value == 'sell_bb_upper4':
                conditions.append(dataframe['close'] < dataframe['sell_bb_upper4'])

        # Check that the candle had volume
        conditions.append(dataframe['volume'] > 0)

        if conditions:
            dataframe.loc[
                reduce(lambda x, y: x & y, conditions),
                'buy'] = 1

        return dataframe

```
When I try to optimize my parameter I get error below:

`2021-08-25 00:27:53,517 - freqtrade - ERROR - The 'buy' space is included into the hyperoptimization but indicator_space() method is not found in your custom Hyperopt class. You should either implement this method or remove the 'buy' space from hyperoptimization.`

I have added buy_*, sell_* prefixes, even added space='buy', space='sell' but still it doesn't work. 

## To replicate the problem:

I used command below:
`freqtrade hyperopt --config config.json --hyperopt-loss SharpeHyperOptLoss --strategy MyFirstStrategy -e 500 --spaces all`

**Comment by xmatthias** [comment]
you've got some key-errors in your sell-side (e.g. `conditions.append(dataframe['rsi'] > self.sell_rsi.value)` - where `'rsi'` is never assigned) - but that's not the only one.

Apart from these errors, this is not explaining the problem you're encountering (these are followup problems that can be resolved easily).

It appears like it's not working with `--spaces all` - but works fine with `--spaces buy sell` ...
I'll see to have that fixed soon.

**Comment by xmatthias** [comment]
The problem was the message, which was slightly wrong / missleading.
it not the buy space  that's missing, but the protection space.
Since you're using `all` as space, the assumption is that you'll be using all available spaces, which are `buy,sell,roi,stoploss,trailing,protection`.
Possible values include (on top of the above) also `default` (which is the default and does not include trailing and protection) - as well as `all`- which includes all of the above (but also assumes ALL are really defined).

This is also explained [here](https://www.freqtrade.io/en/stable/hyperopt/#running-hyperopt-with-smaller-search-space).

**Comment by Heraklaitos** [comment]
@xmatthias you're really great! What you mentioned solved the issue. The error message is indeed misleading but still I wonder I didn't try different space settings. By the way, sorry for all those wrong keys, I have played on the code so much that there remained so many errors :)
Thanks for your really quick response by the way!
